client/src/send_and_confirm_transactions_in_parallel.rs

```diff
@@ -0,0 +1,476 @@
+                            && current_blockheight > x.last_valid_blockheight;
+                        is_not_expired || is_recently_expired
+                    })
+                    .map(|x| *x.key())
+                    .collect();
+                for signatures in
+                    transactions_to_verify.chunks(MAX_GET_SIGNATURE_STATUSES_QUERY_ITEMS)
+                {
+                    if let Ok(result) = rpc_client.get_signature_statuses(signatures).await {
+                        let statuses = result.value;
+                        for (signature, status) in signatures.iter().zip(statuses.into_iter()) {
+                            if let Some((status, data)) = status
+                                .filter(|status| {
+                                    status.satisfies_commitment(rpc_client.commitment())
+                                })
+                                .and_then(|status| {
+                                    unconfirmed_transasction_map
+                                        .remove(signature)
+                                        .map(|(_, data)| (status, data))
+                                })
+                            {
+                                num_confirmed_transactions.fetch_add(1, Ordering::Relaxed);
+                                if let Some(error) = status.err {
+                                    errors_map.insert(data.index, error);
+                                }
+                            };
+                        }
+                    }
+                }
+
+                last_block_height = current_blockheight;
+            }
+            tokio::time::sleep(Duration::from_secs(1)).await;
+        }
+    })
+}
+
+#[derive(Clone, Debug)]
+struct SendingContext {
+    unconfirmed_transasction_map: Arc<DashMap<Signature, TransactionData>>,
+    error_map: Arc<DashMap<usize, TransactionError>>,
+    blockhash_data_rw: Arc<RwLock<BlockHashData>>,
+    num_confirmed_transactions: Arc<AtomicU32>,
+    total_transactions: usize,
+    current_blockheight: Arc<AtomicU64>,
+}
+
+async fn sign_all_messages_and_send<T: Signers + ?Sized>(
+    progress_bar: &Option<indicatif::ProgressBar>,
+    rpc_client: Arc<RpcClient>,
+    tpu_client: &Option<QuicTpuClient>,
+    messages_with_index: Vec<(usize, Message)>,
+    signers: &T,
+    context: &SendingContext,
+) -> Result<()> {
+    let current_transaction_count = messages_with_index.len();
+    // send all the transaction messages
+    for (counter, (index, message)) in messages_with_index.iter().enumerate() {
+        let mut transaction = Transaction::new_unsigned(message.clone());
+        let blockhashdata = *context.blockhash_data_rw.read().await;
+
+        // we have already checked if all transactions are signable.
+        transaction
+            .try_sign(signers, blockhashdata.blockhash)
+            .expect("Transaction should be signable");
+        let signature = transaction.signatures[0];
```

Comment:
maybe we could move this above `context.unconfirmed_transasction_map.insert` so that the declaration and usage at closer to each other (if borrow checker allows)?